### PR TITLE
An email scraper

### PR DESCRIPTION
**Pull request by michaelbjames**
For setting up:
1. `$ virtualenv env`
2. `$ source env/bin/activate`
3. `(env)$ pip install -r requirements.txt`
4. Get the emails of a website!
5. `(env)$ python main.py schlaikjer.net`

Cool things:
1. It finds one more email address on `jana.com` than the challenge suggests! 😸 
2. If you look at your nginx logs, you'll see my email address: best to be a nice spider and tell the server who's poking around

Known limitations:
1. Currently only looks for emails that are marked by a `mailto:` href, so if you just drop an email address on a website, this scraper will not find it. However, this point goes into a broader bucket about how 'smart' we want this scraper to be. It could also try to parse `hello at michaelbjames dot com` but that may be beyond the scope of a simple scraper. What if the email address is surrounded by other text? Do we accept `myemail?hello@michaelbjames.com!my phone numer?...`? I feel that a `mailto:` is the safest way to grab an email address.
2. It's lacking on tests. I don't have 100% coverage. The one tested function gave me the most trouble while developing, so I made sure it works as I expect. I could spend more time on writing tests. If y'all want to see more tests, I'm happy to add more.

I could spend more time sanitizing emails or extracting emails in more nuanced ways (beyond the `mailto:` approach, but it's better to get this out there than try to make it perfect.

### CODE REVIEW

File: crawler.py [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+import scrapy
+from scrapy.linkextractors import LinkExtractor
+from scrapy.spiders import (
+    CrawlSpider,
+    Rule,
+)
+from scrapy.exceptions import DropItem
+
+"""
+A note on Scrapy's architecture:
```
Comment (clee-jana) [inline crawler.py]: thank you for the note

File: crawler.py [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+            follow=True,
+        ),
+    )
+
+    def process_links(self, links):
+        new_links = []
+        domain = self.allowed_domains[0]
+        for link in links:
+            if is_root_domain(domain, link.url):
+                new_links.append(link)
```
Comment (clee-jana) [inline crawler.py]: are we guaranteed the links are unique because of `LinkExtractor(unique=True)`?

File: crawler.py [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+        for addr in email_addrs:
+            email = EmailItem()
+            email['email'] = addr
+            email_items.append(email)
+        return email_items
+
+
+class EmailPipeline(object):
+    def __init__(self):
+        self.emails_seen = set()
```
Comment (clee-jana) [inline crawler.py]: i like the use of the set here

File: crawler.py [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+
+    def process_item(self, item, spider):
+        email = item['email']
+        if email in self.emails_seen:
+            raise DropItem("Duplicate item found: %s" % item)
```
Comment (clee-jana) [inline crawler.py]: where does this exception get handled? why choose to raise something instead of just moving on?

File: test_utils.py [hunk 5]
```diff
@@ -0,0 +1,18 @@
+import unittest
+import crawler_utils
+
+class TestCrawlerUtils(unittest.TestCase):
+
+    def test_is_root_domain(self):
```
Comment (clee-jana) [inline test_utils.py]: 👍  for having any tests

File: crawler.py (hunk 4 above)
Comment (michaelbjames) [inline crawler.py]: It's handled by the pipeline machinery of scrapy. It's possible to have several ItemPipelines connected together, so pipeline A's return is passed to pipeline B. According to the documentation, a pipeline must return a dict with data, an Item, a [Twisted Deferred](https://twistedmatrix.com/documents/current/core/howto/defer.html), or raise a DropItem exception to filter them out of further processing.

File: crawler.py (hunk 2 above)
Comment (michaelbjames) [inline crawler.py]: It guarantees that the links we get _per page_ are unique, so the list of links will be a little smaller. Scrapy makes sure we don't visit the same page twice though. It's a small optimization, and not needed to make the program work as expected. I'm happy to remove it to pair the code down to the essential bits required to run as expected.

**Pull request closed without merging**